openmw: 0.48.0 -> 0.49.0

SigmaSquadron opened the pull request:
Updates OpenMW to the latest major version, 0.49. This also required an update to MyGUI, so I ended up adopting that package. It also adds `collada-dom` back to Nixpkgs, as it is now a required dependency for OpenMW.

Closes #422453, closes #314030, closes #339058, reverts #290366, closes #326874, closes #410968, closes #434886.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [Nixpkgs 25.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/doc/release-notes/rl-2511.section.md) (or backporting [25.05](https://github.com/NixOS/nixpkgs/blob/master/doc/manual/release-notes/rl-2505.section.md) Nixpkgs Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
- [NixOS 25.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2511.section.md) (or backporting [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) NixOS Release notes)
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md), [pkgs/README.md](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md), [maintainers/README.md](https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md) and other contributing documentation in corresponding paths.

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

SigmaSquadron committed:
openmw: 0.48.0 -> 0.49.0

Signed-off-by: Fernando Rodrigues <alpha@sigmasquadron.net>

SigmaSquadron committed:
mygui: 3.4.2 -> 3.4.3

We're now using the Darwin framework patch from OpenMW instead of vendoring it here.
It's the same patch, but fetchpatch saves us a few bytes.

Signed-off-by: Fernando Rodrigues <alpha@sigmasquadron.net>

SigmaSquadron committed:
mygui: refactor

Removes a wide `with` scope, replaces `rec` with `finalAttrs`, adopts the package and adds a `meta.changelog` entry.

Signed-off-by: Fernando Rodrigues <alpha@sigmasquadron.net>

SigmaSquadron committed:
openmw: refactor

The usual code modernisation (finalAttrs, no with lib; in meta, no pname interpolation), and CMAKE variables now use lib.cmake*.

Signed-off-by: Fernando Rodrigues <alpha@sigmasquadron.net>

SigmaSquadron commented:
cc @marius851000. Please request a new invite to the NixOS GitHub org so you can be requested for reviews.

eljamm commented on pkgs/by-name/my/mygui/package.nix:
```diff
@@ -63,10 +68,12 @@ stdenv.mkDerivation rec {
     "-DMYGUI_RENDERSYSTEM=${renderSystem}"
```
See:

- https://wiki.openmw.org/index.php?title=Development_Environment_Setup#MyGUI
- https://gitlab.com/OpenMW/openmw/-/issues/7717

```suggestion
    "-DMYGUI_RENDERSYSTEM=${renderSystem}"
    "-DMYGUI_DONT_USE_OBSOLETE=ON"
```

eljamm commented on pkgs/games/openmw/default.nix:
```diff
@@ -20,29 +19,34 @@
   recastnavigation,
```
```suggestion
  qttools,
  recastnavigation,
```

eljamm commented on pkgs/games/openmw/default.nix:
```diff
@@ -99,7 +101,7 @@ stdenv.mkDerivation rec {
     boost
     bullet'
-    ffmpeg_6
+    ffmpeg
```
```suggestion
    ffmpeg
    qttools
```

eljamm reviewed (commented):
Btw, this PR also closes https://github.com/NixOS/nixpkgs/pull/314030

SigmaSquadron commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 422562`
Commit: `d94191894b66dbc941768633c0526f54c14971ab`

---
### `x86_64-linux`
<details>
  <summary>:x: 2 packages failed to build:</summary>
  <ul>
    <li>openmw-tes3mp</li>
    <li>stuntrally</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 4 packages built:</summary>
  <ul>
    <li>mygui</li>
    <li>openmw</li>
    <li>portmod</li>
    <li>portmod.dist</li>
  </ul>
</details>

---
<details>
<summary>Error logs: `x86_64-linux`</summary>
<details>
<summary>stuntrally</summary>
<pre>      |                                                           ^~~~~~~~~~~~
/nix/store/iijp6lasnaszvcpdj5k2m9a9w5dgmclw-mygui-3.4.3/include/MYGUI/MyGUI_Delegate.h:331:23: note: declared here
  331 |                 using CMultiDelegate1 MYGUI_OBSOLETE("use : MyGUI::delegates::MultiDelegate") = MultiDelegate<Args...>;
      |                       ^~~~~~~~~~~~~~~
In file included from /build/source/source/ogre/common/MessageBox/MessageBox.h:13:
/build/source/source/ogre/common/MessageBox/BaseLayout.h: In instantiation of 'T* wraps::BaseLayout::_createFakeWidget(MyGUI::Widget*) [with T = MyGUI::Widget]':
/build/source/source/ogre/common/MessageBox/BaseLayout.h:124:53:   required from here
  124 |                                                 mMainWidget = _createFakeWidget<MyGUI::Widget>(_parent);
      |                                                               ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~
/build/source/source/ogre/common/MessageBox/BaseLayout.h:212:86: error: cannot convert 'std::string_view' {aka 'std::basic_string_view<char>'} to 'const std::string&' {aka 'const std::__cxx11::basic_string<char>&'}
  212 |                         return static_cast<T*>(_createFakeWidgetT(T::getClassTypeName(), _parent));
      |                                                                   ~~~~~~~~~~~~~~~~~~~^~
      |                                                                                      |
      |                                                                                      std::string_view {aka std::basic_string_view<char>}
/build/source/source/ogre/common/MessageBox/BaseLayout.h:215:70: note:   initializing argument 1 of 'MyGUI::Widget* wraps::BaseLayout::_createFakeWidgetT(const std::string&, MyGUI::Widget*)'
  215 |                 MyGUI::Widget* _createFakeWidgetT(const std::string& _typeName, MyGUI::Widget* _parent)
      |                                                   ~~~~~~~~~~~~~~~~~~~^~~~~~~~~
make[2]: *** [source/CMakeFiles/stuntrally.dir/build.make:208: source/CMakeFiles/stuntrally.dir/ogre/CHud.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:322: source/CMakeFiles/stuntrally.dir/all] Error 2
make: *** [Makefile:136: all] Error 2</pre>
</details>
</details>

eljamm commented:
Can't test as I'm encountering this issue: https://github.com/NixOS/nixpkgs/issues/326874

qubitnano commented:
Builds and loads for me by reverting [290366](https://github.com/NixOS/nixpkgs/pull/290366) and patching for boost

```
From d3a2949cc8022a2c7406f041d69b109fb6a2d53c Mon Sep 17 00:00:00 2001
From: qubitnano <146656568+qubitnano@users.noreply.github.com>
Date: Sat, 5 Jul 2025 16:48:06 -0400
Subject: [PATCH] collada-dom: reinit at 0-unstable-2020-01-03

---
 pkgs/by-name/co/collada-dom/package.nix    | 54 ++++++++++++++++++++++
 pkgs/by-name/op/openscenegraph/package.nix |  4 +-
 pkgs/top-level/aliases.nix                 |  1 -
 3 files changed, 56 insertions(+), 3 deletions(-)
 create mode 100644 pkgs/by-name/co/collada-dom/package.nix

diff --git a/pkgs/by-name/co/collada-dom/package.nix b/pkgs/by-name/co/collada-dom/package.nix
new file mode 100644
index 000000000000..fa790e5aaacd
--- /dev/null
+++ b/pkgs/by-name/co/collada-dom/package.nix
@@ -0,0 +1,54 @@
+{
+  stdenv,
+  lib,
+  fetchFromGitHub,
+  cmake,
+  boost,
+  libxml2,
+  minizip,
+  readline,
+  fetchpatch,
+}:
+
+stdenv.mkDerivation {
+  pname = "collada-dom";
+  version = "0-unstable-2020-01-03";
+
+  src = fetchFromGitHub {
+    owner = "rdiankov";
+    repo = "collada-dom";
+    rev = "c1e20b7d6ff806237030fe82f126cb86d661f063";
+    hash = "sha256-A1ne/D6S0shwCzb9spd1MoSt/238HWA8dvgd+DC9cXc=";
+  };
+
+  patches = [
+    # Fix building on boost >=1.85.0
+    (fetchpatch {
+      name = "boost1.85.patch";
+      url = "https://github.com/rdiankov/collada-dom/pull/43.patch";
+      hash = "sha256-8aSc21IaBdvhX8QKhBt8mzy/WfinRMI22ZlH9M0QJKg=";
+    })
+  ];
+
+  postInstall = ''
+    chmod +w -R $out
+    ln -s $out/include/*/* $out/include
+  '';
+
+  nativeBuildInputs = [ cmake ];
+
+  buildInputs = [
+    boost
+    libxml2
+    minizip
+    readline
+  ];
+
+  meta = with lib; {
+    description = "Lightweight version of collada-dom, with only the parser.";
+    homepage = "https://github.com/rdiankov/collada-dom";
+    license = licenses.mit;
+    maintainers = with maintainers; [ marius851000 ];
+    platforms = platforms.all;
+  };
+}
diff --git a/pkgs/by-name/op/openscenegraph/package.nix b/pkgs/by-name/op/openscenegraph/package.nix
index e4c12742965b..9ae2b719dab2 100644
--- a/pkgs/by-name/op/openscenegraph/package.nix
+++ b/pkgs/by-name/op/openscenegraph/package.nix
@@ -32,7 +32,7 @@
   curlSupport ? true,
   curl,
   colladaSupport ? false,
-  opencollada,
+  collada-dom,
   opencascadeSupport ? false,
   opencascade-occt,
   ffmpegSupport ? false,
@@ -98,7 +98,7 @@ stdenv.mkDerivation rec {
     ++ lib.optional gdalSupport gdal
     ++ lib.optional curlSupport curl
     ++ lib.optionals colladaSupport [
-      opencollada
+      collada-dom
       pcre
     ]
     ++ lib.optional opencascadeSupport opencascade-occt
diff --git a/pkgs/top-level/aliases.nix b/pkgs/top-level/aliases.nix
index c69ff6e49ecf..ac7a3b5e9838 100644
--- a/pkgs/top-level/aliases.nix
+++ b/pkgs/top-level/aliases.nix
@@ -449,7 +449,6 @@ mapAliases {
   cloudlogoffline = throw "cloudlogoffline has been removed"; # added 2025-05-18
   clwrapperFunction = throw "Lisp packages have been redesigned. See 'lisp-modules' in the nixpkgs manual."; # Added 2024-05-07
   CoinMP = coinmp; # Added 2024-06-12
-  collada-dom = opencollada; # added 2024-02-21
   colorpicker = throw "'colorpicker' has been removed due to lack of maintenance upstream. Consider using 'xcolor', 'gcolor3', 'eyedropper' or 'gpick' instead"; # Added 2024-10-19
   colorstorm = throw "'colorstorm' has been removed because it was unmaintained in nixpkgs and upstream was rewritten."; # Added 2025-06-15
   connman-ncurses = throw "'connman-ncurses' has been removed due to lack of maintenance upstream."; # Added 2025-05-27
-- 
2.49.0
```

SigmaSquadron commented:
I guess the way forward is to re-add `collada-dom` then.

@K900: Would it be alright to add `collada-dom` to Nixpkgs again with the necessary build patches?

marius851000 commented:
Interesting. I don’t remember the lack collada causing a fatal error on 0.48. I have this PR ( #339058 ), if you need some help into that. I’ll be happy if you include the work there into this PR. I can also provide a mod using those DAE model if needed.

SigmaSquadron committed:
collada-dom: init at 2.5.1

Previously removed due to being unmaintained, this commit
adds collada-dom back to Nixpkgs, as opencollada proved an
insufficient replacement to it. The new upstream is slightly
more modern than the previous one.

Co-authored-by: Marius David <marius@mariusdavid.fr>
Signed-off-by: Fernando Rodrigues <alpha@sigmasquadron.net>

SigmaSquadron committed:
openscenegraph: fix collada implementation

`opencollada` is not the same as `collada-dom`. OpenSceneGraph requires
`collada-dom` to generate some shared object files required by OpenMW.

Co-authored-by: Marius David <marius@mariusdavid.fr>
Signed-off-by: Fernando Rodrigues <alpha@sigmasquadron.net>

SigmaSquadron commented:
it works with `collada-dom`, and the `osgdb_dae.so` file is correctly generated :)

@marius851000: I've used your commits, but couldn't figure out how to preserve the authorship information when I edited them slightly. I've added your information to a `Co-authored-by` tag, which I hope is appropriate.

![works](https://github.com/user-attachments/assets/5e929ac3-dada-43dd-8f78-5b05dd394957)

marius851000 commented:
Interesting. Portmod build on my side (x86-64 linux). It looks like the way the sandbox is set up does not allow portmod to use its own sandboxing mechanism. It work on my machine, which does have the nix sandbox enabled.

A solution to that may be to disable the failing tests, but that’s about 1/4th of them.

(and otherwise, I’m fine with how you credited me in the re-used commit. It’s not something I bring a lot of attention to anyway)

marius851000 commented:
As far as tes3mp is concerned, however, it looks like this custom OSG is causing issues. I haven’t tested, but something along the line of (in `openmw.nix`) `(openmw.override { openscenegraph = openscenegraph; }).overrideAttrs` might do the trick.

peterhoeg commented on pkgs/by-name/co/collada-dom/package.nix:
```diff
@@ -0,0 +1,54 @@
+    owner = "rdiankov";
+    repo = "collada-dom";
+    rev = "c1e20b7d6ff806237030fe82f126cb86d661f063";
+    sha256 = "sha256-A1ne/D6S0shwCzb9spd1MoSt/238HWA8dvgd+DC9cXc=";
```
sha256 -> hash

peterhoeg commented on pkgs/games/openmw/default.nix:
```diff
@@ -3,46 +3,52 @@
-        url = "https://gitlab.com/OpenMW/openmw-dep/-/raw/0abe3c9c3858211028d881d7706813d606335f72/macos/osg.patch";
-        sha256 = "sha256-/CLRZofZHot8juH78VG1/qhTHPhy5DoPMN+oH8hC58U=";
+        url = "https://gitlab.com/OpenMW/openmw-dep/-/raw/1305497c009dc0e7a6a70fe14f0a2f92b96cbcb4/macos/osg.patch";
+        sha256 = "sha256-G8Y+fnR6FRGxECWrei/Ixch3A3PkRfH6b5q9iawsSCY=";
```
sha256 -> hash

peterhoeg commented on pkgs/games/openmw/default.nix:
```diff
@@ -3,46 +3,52 @@
+  wrapQtAppsHook,
   yaml-cpp,
+
+  withGL ? "GLVND",
```
`withXXX` are *typically* used for boolean flags. I suggest something like `GLPreference`

peterhoeg commented on pkgs/games/openmw/default.nix:
```diff
@@ -99,35 +103,34 @@ stdenv.mkDerivation rec {
+  cmakeFlags = [
+    (lib.cmakeFeature "OpenGL_GL_PREFERENCE" withGL)
+    (lib.cmakeBool "OPENMW_USE_SYSTEM_RECASTNAVIGATION" true)
+  ] ++ lib.optional stdenv.hostPlatform.isDarwin (lib.cmakeBool "OPENMW_OSX_DEPLOYMENT" true);
```
I guess you could just do `(lib.cmakeBool "OPENMW_OSX_DEPLOYMENT" stdenv.hostPlatform.isDarwin)` and drop the optional bit. It reads better.

peterhoeg commented on pkgs/games/openmw/default.nix:
```diff
@@ -99,35 +103,34 @@ stdenv.mkDerivation rec {
+      sigmasquadron
     ];
-    platforms = platforms.linux ++ platforms.darwin;
+    platforms = with lib.platforms; (linux ++ darwin);
```
`lib.platforms.unix` should do the trick.

peterhoeg reviewed (dismissed):
Other than the small things mentioned here, there is also:

1. with the move to `pkgs/by-name`, you need to drop the reference in `pkgs/top-level/all-packages.nix`
2. with the move to `pkgs/by-name`, you need to pass in `qt6Packages` and drop the main `qttools, wrapQtAppsHook` et al arguments.
3. fix merge conflicts

SigmaSquadron commented:
Sorry for abandoning this, and thank you for the review! All changes should be addressed now, except for the Darwin failures and the portmod transient failure/tes3mp failures. Those seem above my pay grade tbh.

SigmaSquadron commented on pkgs/by-name/op/openmw/package.nix:
```diff
@@ -0,0 +1,135 @@
+    unshield
+    yaml-cpp
+  ]
+  ++ (with qt6Packages; [ qttools ]);
```
Nope, I am definitely not sure, but I'm now rebuilding with `strictDeps` enabled, which should iron out these issues.

SigmaSquadron replied:
Looks like that's correct; `qttools` is not executed during the build, so it goes to `buildInputs`.

peterhoeg commented:
The changes to `openscenegraph` breaks the build. I'll push some fixes to your branch tomorrow.

SigmaSquadron commented:
I don't see how? I built osg successfully for OpenMW, and the non-openmw copy should be unaffected.

SigmaSquadron commented:
Ah! I didn't think `null` would set it to anything. Easy enough fix though.

nim65s commented:
fwiw I have a fork of collada-dom with a bit more PR merged: https://github.com/Gepetto/collada-dom

SigmaSquadron committed:
openmw-tes3mp: drop

The multiplayer fork of OpenMW is currently unmaintained,
fails to build with OpenMW 0.49.0, and needs multiple GCC
14 compilation patches.

Signed-off-by: Fernando Rodrigues <alpha@sigmasquadron.net>

marcin-serwin reviewed (changes requested):
It builds but doesn't work for me on `aarch64-linux`, it starts and shows Bethesda logo but crashes before reaching menu with this error:
```
*** Fatal Error ***
Address not mapped to object (signal 11)
Address: 0x7fff08929000
```
The 0.48 version from `master` works correctly.

SigmaSquadron commented on pkgs/by-name/co/collada-dom/package.nix:
```diff
@@ -0,0 +1,52 @@
+  pname = "collada-dom";
+  version = "2.5.1";
+
+  __structuredAttrs = true;
```
`__structuredAttrs` is a recent recommendation for package refactoring, as it will eventually be enabled by default for all derivations. It's good to enable it early and see if anything breaks.

SigmaSquadron commented on pkgs/by-name/op/openmw/package.nix:
```diff
@@ -0,0 +1,143 @@
+      marius851000
+      sigmasquadron
+    ];
+    platforms = lib.platforms.unix;
```
https://github.com/NixOS/nixpkgs/pull/422562#discussion_r2250678245
I'm not sure what the best solution here is. We could have `linux ++ darwin ++ windows` && `badPlatforms = windows`

SigmaSquadron replied:
Should be improved now.

SigmaSquadron commented:
This is ready for a final review. I'd recommend going commit-by-commit.

SigmaSquadron commented:
I went for a more generic isAarch64Linux approach, but the result is the same.

nim65s reviewed (approved):
I'm happy to report that with this, gepetto-gui is able to display robots with collada meshes, eg:

- switch to this PR
- [remove qtwebengine from pythonqt to fix it](https://github.com/NixOS/nixpkgs/pull/437521)
- replace `openscenegraph` with `openscenegraph.override { colladaSupport = true; }` in gepetto-viewer as it is done in openmw
- `nix-shell -I nixpkgs=. -p 'python3.withPackages(p: [p.example-robot-data p.gepetto-gui])'`
- run `gepetto-gui &` and `python -m example_robot_data romeo` (talos works without collada-dom, but not romeo)

Thanks for the work !

marcin-serwin reviewed (approved):
Builds and runs on `aarch64-linux`.

SigmaSquadron commented:
cc @pSub: Looks like this breaks `stuntrally`. Could you take a look here?

iedame commented:
> cc @pSub: Looks like this breaks `stuntrally`. Could you take a look here?

Looks like stuntrally is seriously outdated, version `2.7` - from `11. 11. 2022` https://github.com/NixOS/nixpkgs/issues/416166
2.x changelog: https://github.com/stuntrally/stuntrally/blob/master/docs/changelog.md
recent changelog: https://github.com/stuntrally/stuntrally3/blob/main/docs/Changelog.md

I'm not sure the package is maintained anymore, last maintainer commit is from 1189 days ago.

SigmaSquadron commented:
oh, lmao
